[Chau Le] iP

ChauuuLe opened the pull request:
# 🚀 Dak - Your Personal Task Assistant  

> **"Your mind is for having ideas, not holding them."** – David Allen  

## 📌 About Dak  
Dak is a simple yet powerful **task management chatbot** that helps you organize your tasks efficiently.  
It keeps track of your to-dos, deadlines, and events, so you don't have to!  

### ✨ Features  
- ✅ **Task management** (`Todo`, `Deadline`, `Event`)  
- ✅ **Marking & Unmarking tasks** (`mark`, `unmark`)  
- ✅ **Task deletion** (`delete`)  
- ✅ **Task searching** (`find`)  
- ✅ **Sorting tasks by deadline** (`sort`)  
- 🔄 **Persistent data storage** (`save` & `load`)  

## 🔢 How to Use Dak  
1. **Download** Dak from [this link](https://github.com/ChauuuLe/ip).  
2. **Double-click** the file to open it.  
3. **Start managing tasks** by adding them using commands like `todo Read Book`, `deadline Submit Report /by 10/3/2024`, etc.  
4. **Let Dak handle your tasks for you!** 😉  

## 📝 Code Example  
If you are a Java programmer, you can use Dak to practice Java too! Here's the **`main()` method**:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(dak.ui.MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename project

Commit:
Add greeting and goodbye message

Commit:
Change greeting and bye, add echo

Commit:
Add list item

Commit:
Add mark as done

Commit:
Add deadline and event and todo for tasks

Commit:
Add git ignore .class files

Commit:
Add handle exception error

Commit:
Add delete command

Commit:
Add text ui testing

Commit:
Add Storage class for making storage

Commit:
Add java doc for files in tasks package

Commit:
Add save data to chatbot

Commit:
Merge branch 'branch-Level-7' into master

Commit:
Change date and time format

Commit:
Merge branch 'branch-Level-8' into master

Commit:
Change OOP

Commit:
Change package structure

nja99 commented on src/main/java/dak/storage/Storage.java:
```diff
@@ -0,0 +1,87 @@
+                return event;
+            default:
+                throw new IllegalArgumentException("Invalid task type in file: " + taskType);
+        }
```
Should the case clauses be indented?

nja99 commented on src/main/java/dak/task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+
+    // Define the custom date-time format
+    private static final DateTimeFormatter INPUT_FORMAT = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+    private static final DateTimeFormatter OUTPUT_FORMAT = DateTimeFormatter.ofPattern("MMM d yyyy, h:mm a");
```
Would it be better to extract the date time formatting into its own utility class?

nja99 commented on src/main/java/dak/task/TaskList.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @param ui The Ui object to handle printing.
+     */
+    public void listTasks(Ui ui) {
+        if (tasks.isEmpty()) {
+            ui.printMessage("You have no tasks in your list.");
+        } else {
+            StringBuilder listMessage = new StringBuilder("Here are the tasks in your list:");
+            for (int i = 0; i < tasks.size(); i++) {
+                listMessage.append("\n  ").append(i + 1).append(". ").append(tasks.get(i));
+            }
+            ui.printMessage(listMessage.toString());
+        }
+    }
```
Is there any reason why you chose to print the message here rather than returning a string to be handled at a higher level?

nja99 commented on src/main/java/dak/exceptions/DukeException.java:
Perhaps refactor this to align with your chatbot name?
Also would it be better to have more specific exceptions?

nja99 reviewed (commented):
Overall, the code was easy to comprehend with only minor coding standard violations.

SAN-MUYUN commented on src/main/java/dak/command/AddCommand.java:
```diff
@@ -0,0 +1,93 @@
+            if (timeParts.length < 2) {
+                throw new DukeException("Invalid format. Use: event <description> /from <d/M/yyyy HHmm> /to <d/M/yyyy HHmm>");
+            }
+            String from = timeParts[0].trim();
```
might be good to change variable names ```from```  to ```startTime ``` and ``` to``` to ```endTime```.

SAN-MUYUN commented on src/main/java/dak/parser/Parser.java:
```diff
@@ -0,0 +1,50 @@
+package dak.parser;
+
+import dak.command.*;
```
It may not be a good idea to import using wildcard ```*```.

SAN-MUYUN commented on src/main/java/dak/storage/Storage.java:
```diff
@@ -0,0 +1,87 @@
+import dak.task.Todo;
+import dak.task.Deadline;
+import dak.task.Event;
+import java.io.*;
```
Same here, it would be better to import individual classes instead of using the wildcard ```*```.

SAN-MUYUN commented on src/main/java/dak/task/TaskList.java:
```diff
@@ -0,0 +1,62 @@
+            ui.printMessage("You have no tasks in your list.");
+        } else {
+            StringBuilder listMessage = new StringBuilder("Here are the tasks in your list:");
+            for (int i = 0; i < tasks.size(); i++) {
```
good naming of variables, making the code very easy to read and understand. 👍

SAN-MUYUN commented on src/main/java/dak/ui/Ui.java:
```diff
@@ -0,0 +1,74 @@
+    public void printMessage(String message) {
+        System.out.println("  ____________________________________________________________");
+        System.out.println("  " + message);
+        System.out.println("  ____________________________________________________________");
```
Perhaps it is a good idea to reuse the ```showLine()``` method defined below?

SAN-MUYUN reviewed (commented):
Overall LGTM, with a few nits to be fixed. 👍 👍 👍

Commit:
Add missing java doc

Commit:
Fix coding standard

Commit:
Merge branch 'A-JavaDoc'

Commit:
Merge branch 'A-JavaDoc'

Commit:
Java GUI

Commit:
Java GUI

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Add gradle support

Commit:
Add JUnit tests

Commit:
Add find command

Commit:
Merge branch 'branch-Level-9'

Commit:
Fix import error

Commit:
Merge branch 'master' into branch-Level-10

Commit:
Add javaFX to the chatbot

Commit:
Refactor tests for command classes

Commit:
Add assertions to MarkCommand, UnmarkCommand, FindCommand and Parser for robustness

ChauuuLe committed:
Merge A-Assertions branch into master

Add assertions to improve robustness in MarkCommand, UnmarkCommand, FindCommand, and Parser.

- Ensures task index is valid before execution in MarkCommand and UnmarkCommand.
- Adds null checks for UI, Storage, and TaskList to prevent unexpected crashes.
- Verifies search queries are not null or empty in FindCommand.

Commit:
Merge branch 'master' of https://github.com/ChauuuLe/ip into branch-A-CodeQuality

Commit:
Fix Javadoc formatting for consistency

Commit:
Refactor Parser to improve readability and maintainability

The 'parse()' method in 'Parser' is too long and difficult to maintain, as it handles multiple command types within a single block of code. This violates the Single Level of Abstraction Principle (SLAP), making it harder to read and modify.

To improve readability and maintainability:
- Extracted simple commands ('bye', 'list') handling into 'parseSimpleCommand()'.
- Extracted commands requiring a task number ('mark', 'unmark', 'delete') into 'parseTaskNumberCommand()'.
- Extracted add commands ('todo', 'deadline', 'event') into 'parseAddCommand()'.
- Extracted find command handling into 'parseFindCommand()'.
- Introduced helper methods ('isSimpleCommand()', 'isTaskNumberCommand()', 'isAddCommand()')   to determine command types before parsing.

With these changes, 'parse()' now follows SLAP and is easier to read, test, and extend.

ChauuuLe committed:
Merge branch-A-CodeQuality into master

Refactor to improve code quality

Commit:
Add sorting functionality for deadlines

Commit:
Add image for avatars of chatbot and user

Commit:
Add FXML styling for improved GUI

Commit:
Add error handling for event and deadline creation

Commit:
Fix errors of sort command

Commit:
Add user guide for chat bot

Commit:
Add more test for all commands

Commit:
Add missing javadoc
